code/modules/atmospherics/machinery/components/machines/crystallizer.dm

```diff
@@ -0,0 +1,341 @@
+		nodes[2] = null
+
+	if(parents[1])
+		nullifyPipenet(parents[1])
+	if(parents[2])
+		nullifyPipenet(parents[2])
+
+	atmosinit()
+	node1 = nodes[1]
+	if(node1)
+		node1.atmosinit()
+		node1.addMember(src)
+	node2 = nodes[2]
+	if(node2)
+		node2.atmosinit()
+		node2.addMember(src)
+	SSair.add_to_rebuild_queue(src)
+	return TRUE
+
+/obj/machinery/atmospherics/components/binary/crystallizer/update_icon()
+	. = ..()
+	cut_overlays()
+
+	if(panel_open)
+		icon_state = "[base_icon]-open"
+	else if(on && is_operational)
+		icon_state = "[base_icon]-on"
+	else
+		icon_state = "[base_icon]-off"
+
+	add_overlay(getpipeimage(icon, "pipe", dir, COLOR_LIME, piping_layer))
+	add_overlay(getpipeimage(icon, "pipe", turn(dir, 180), COLOR_MOSTLY_PURE_RED, piping_layer))
+
+/obj/machinery/atmospherics/components/binary/crystallizer/attackby_secondary(mob/user)
+	if(!can_interact(user))
+		return
+	on = !on
+	investigate_log("was turned [on ? "on" : "off"] by [key_name(user)]", INVESTIGATE_ATMOS)
+	update_icon()
+
+///Checks if the gases in the input are the ones needed by the recipe
+/obj/machinery/atmospherics/components/binary/crystallizer/proc/check_gas_requirements()
+	var/datum/gas_mixture/contents = airs[2]
+	var/list/recipe = GLOB.gas_recipe_meta[recipe_type]
+	for(var/gas_type in recipe[META_RECIPE_REQUIREMENTS])
+		if(!contents.gases[gas_type] || !contents.gases[gas_type][MOLES])
+			return FALSE
+	return TRUE
+
+///Checks if the reaction temperature is inside the range of temperature + a little deviation
+/obj/machinery/atmospherics/components/binary/crystallizer/proc/check_temp_requirements()
+	var/datum/gas_mixture/contents = airs[2]
+	var/list/recipe = GLOB.gas_recipe_meta[recipe_type]
+	if(contents.temperature >= (recipe[META_RECIPE_MIN_TEMP] * MIN_DEVIATION_RATE) && contents.temperature <= (recipe[META_RECIPE_MAX_TEMP] * MAX_DEVIATION_RATE))
+		return TRUE
+	return FALSE
+
+///Injects the gases from the input inside the internal gasmix, the amount is dependant on the gas_input var
+/obj/machinery/atmospherics/components/binary/crystallizer/proc/inject_gases()
+	var/datum/gas_mixture/contents = airs[2]
+	var/list/recipe = GLOB.gas_recipe_meta[recipe_type]
+	for(var/gas_type in recipe[META_RECIPE_REQUIREMENTS])
+		internal.merge(contents.remove_specific(gas_type, contents.gases[gas_type][MOLES] * (gas_input * 0.1)))
+
+///Checks if the gases required are all inside
+/obj/machinery/atmospherics/components/binary/crystallizer/proc/internal_check()
+	gas_check = 0
+	var/list/recipe = GLOB.gas_recipe_meta[recipe_type]
+	for(var/gas_type in recipe[META_RECIPE_REQUIREMENTS])
+		if(internal.gases[gas_type][MOLES] >= recipe[META_RECIPE_REQUIREMENTS][gas_type])
+			gas_check++
+
+///Calculation for the heat of the various gas mixes and controls the quality of the item
+/obj/machinery/atmospherics/components/binary/crystallizer/proc/heat_calculations()
+	if(airs[1].total_moles() * 0.05 > MINIMUM_MOLE_COUNT)
+		var/datum/gas_mixture/cooling_port = airs[1]
```

Comment:
Why cast airs[1] after the check and not before?

Reply:
uh no idea, will move it